labs/12/theory.tex

```diff
@@ -0,0 +1,106 @@
+First, define a prior length-scale $l$. This captures our belief over the function frequency. A short length-scale $l$ corresponds to high frequency data, and a long length-scale corresponds to low frequency data.Take the length-scale squared, and divide it by the weight decay. We then scale the result by half the dropout probability over the number of data points. Mathematically this result in a Gaussian process precision $\tau = \frac{l^2 p}{2N \lambda}$ we mentioned above. Note that $p$ here is the probability of the units not being dropped - in most implementations $p_{code}$ is defined as the probability of the units to be dropped, thus $p := 1- p_{code}$ should be used when calculating $\tau$. Next, simulate a network output with input $x^*$, treating dropout as if we were using it during training phase. This means enabling dropout also in test phase. Repeat this $T$ times with different units dropped every time, and collect the results ${\hat{y_t}^* (x^*)}$. These are empirical samples from our approximate predictive posterior. We can get an empirical estimator for the predictive mean of our approximate posterior as well as the predictive variance (our uncertainty) from these samples. Mathematically, averaging forward passes through the network is equivalent to Monte Carlo integration over a Gaussian process posterior approximation. The derivation is provided in \cite{gal2016dropout}. The process of applying dropout into the neural network at both training and test phases is thus called Monte-Carlo dropout (MC dropout). We simply follow these two equations:
+\begin{equation}
+    \mathbb{E} (y^*) \approx \frac{1}{T} \sum^{T}_{t=1} \hat{y_t}^*(x^*)
+\end{equation}
+\begin{equation}
+    Var(y^*) \approx \tau^{-1} I_D + \frac{1}{T}\sum^{T}_{t=1} \hat{y_t}^*(x^*)^T \hat{y_t}^* (x^*) - \mathbb{E} (y^*)^T \mathbb{E} (y^*)
+\end{equation}
+\\
+Eq. 17.1 was introduced as model averaging and it was explained that scaling the weights at test time without dropout gives a reasonable approximation to this equation. Eq. 17.2 is simply the sample variance of $T$ forward passes through the network plus the inverse model precision. Note that the vectors above are row vectors and that the products are outer-products.
+\\
+Implementation in python code to get the predictive mean and uncertainty is easy:
+\begin{figure}[H]
+    \centering
+    \includegraphics[width=\textwidth]{labs/12/images/python code of getting mean and var.PNG}
+    \caption{Python code to obtain predictive mean and uncertainty from dropout}
+    \label{fig:visual_caption}
```

Comment:
`visual_caption` is already defined here https://github.com/Atcold/SP19-DL-collaborative-notes/blob/master/labs/07/theory.tex#L14

Reply:
Please check PR Lab12#53